### PR TITLE
Fix: Prevents merging of tool arguments during preprocessing

### PR DESCRIPTION
**Pull request by greenhestu**
# Description

This PR modifies the data preprocessing logic for tool calls. It resolves an issue where arguments from multiple tools were being merged during preprocessing, causing unnecessary fields to be populated with null values.

This change reforms the structure of tool_calls within the preprocessed data. By removing extraneous null arguments, each tool call in the training data now strictly adheres to its own schema, ensuring it matches the format expected by standard chat templates.

### Problem: 
The current logic merges the function signatures of all available tools, likely due to the schema unification behavior of Hugging Face's datasets.load_dataset with JSON files. (See [HF Docs on JSON loading](https://huggingface.co/docs/datasets/loading#json)).
<br>

### Impact: 
This results in training data where tool_calls contain extraneous arguments with null values.
A model fine-tuned on this data performs poorly when used for inference with standard chat templates (e.g., in vLLM), as there is a mismatch between the training format and the expected inference format.
<br>

### Example:
```json
{"messages":[{"role":"user","content":"move to (0, 1)"},{"role":"assistant","content":"","tool_calls":[{"function":{"name":"move","arguments":{"x":0,"y":1}}}]}],"tools":[{"type":"function","function":{"name":"move","description":"Move to a given location measured in meters","parameters":{"type":"object","properties":{"x":{"type":"number","description":"The x coordinate of the location, negative values are to the left, positive values are to the right"},"y":{"type":"number","description":"The y coordinate of the location, negative values are backward, positive values are forward"}},"required":["x","y"]}}},{"type":"function","function":{"name":"turn","description":"Turn the robot to a given direction","parameters":{"type":"object","properties":{"theta":{"type":"integer","description":"The angle to turn to, in degrees, positive values are counter-clockwise, negative values are clockwise"}},"required":["theta"]}}},{"type":"function","function":{"name":"invalid_prompt","description":"call when the user's prompt is invalid","parameters":{"type":"object","properties":{"message":{"type":"string","description":"why the prompt is invalid"}},"required":["message"]}}}],"add_generation_prompt":false}
{"messages":[{"role":"user","content":"turn 270 degree"},{"role":"assistant","content":"","tool_calls":[{"function":{"name":"turn","arguments":{"theta": 270}}}]}],"tools":[{"type":"function","function":{"name":"move","description":"Move to a given location measured in meters","parameters":{"type":"object","properties":{"x":{"type":"number","description":"The x coordinate of the location, negative values are to the left, positive values are to the right"},"y":{"type":"number","description":"The y coordinate of the location, negative values are backward, positive values are forward"}},"required":["x","y"]}}},{"type":"function","function":{"name":"turn","description":"Turn the robot to a given direction","parameters":{"type":"object","properties":{"theta":{"type":"integer","description":"The angle to turn to, in degrees, positive values are counter-clockwise, negative values are clockwise"}},"required":["theta"]}}},{"type":"function","function":{"name":"invalid_prompt","description":"call when the user's prompt is invalid","parameters":{"type":"object","properties":{"message":{"type":"string","description":"why the prompt is invalid"}},"required":["message"]}}}],"add_generation_prompt":false}
{"messages":[{"role":"user","content":"jump high"},{"role":"assistant","content":"","tool_calls":[{"function":{"name":"invalid_prompt","arguments":{"message": "jump is not a valid action"}}}]}],"tools":[{"type":"function","function":{"name":"move","description":"Move to a given location measured in meters","parameters":{"type":"object","properties":{"x":{"type":"number","description":"The x coordinate of the location, negative values are to the left, positive values are to the right"},"y":{"type":"number","description":"The y coordinate of the location, negative values are backward, positive values are forward"}},"required":["x","y"]}}},{"type":"function","function":{"name":"turn","description":"Turn the robot to a given direction","parameters":{"type":"object","properties":{"theta":{"type":"integer","description":"The angle to turn to, in degrees, positive values are counter-clockwise, negative values are clockwise"}},"required":["theta"]}}},{"type":"function","function":{"name":"invalid_prompt","description":"call when the user's prompt is invalid","parameters":{"type":"object","properties":{"message":{"type":"string","description":"why the prompt is invalid"}},"required":["message"]}}}],"add_generation_prompt":false}
```

Applying the chat template to the data above produces the following result (first column)
``` text
<|im_start|>system
# Tools

You may call one or more functions to assist with the user query.

You are provided with function signatures within <tools></tools> XML tags:
<tools>
{"type": "function", "function": {"name": "move", "description": "Move to a given location measured in meters", "parameters": {"type": "object", "properties": {"x": {"type": "number", "description": "The x coordinate of the location, negative values are to the left, positive values are to the right"}, "y": {"type": "number", "description": "The y coordinate of the location, negative values are backward, positive values are forward"}}, "required": ["x", "y"]}}}
{"type": "function", "function": {"name": "turn", "description": "Turn the robot to a given direction", "parameters": {"type": "object", "properties": {"theta": {"type": "integer", "description": "The angle to turn to, in degrees, positive values are counter-clockwise, negative values are clockwise"}}, "required": ["theta"]}}}
{"type": "function", "function": {"name": "invalid_prompt", "description": "call when the user's prompt is invalid", "parameters": {"type": "object", "properties": {"message": {"type": "string", "description": "why the prompt is invalid"}}, "required": ["message"]}}}
</tools>

For each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:
<tool_call>
{"name": <function-name>, "arguments": <args-json-object>}
</tool_call><|im_end|>
<|im_start|>user
move to (0, 1)<|im_end|>
<|im_start|>assistant
<think>

</think>

<tool_call>
{"name": "move", "arguments": {"x": 0, "y": 1}}
</tool_call><|im_end|>
```
<br>

axolotl preprocess --debug makes following output

* Before this PR (contains "theta": null and "message" null)
``` text
(User prompt and tool tokens omitted.)
(-100, 198) <|im_start|>(-100, 151644) assistant(-100, 77091) 
(-100, 198) <think>(-100, 151667) 

(-100, 271) </think>(-100, 151668) 

(-100, 271) <tool_call>(151657, 151657) 
(198, 198) {"(4913, 4913) name(606, 606) ":(788, 788)  "(330, 330) move(3397, 3397) ",(497, 497)  "(330, 330) arguments(16370, 16370) ":(788, 788)  {"(5212, 5212) x(87, 87) ":(788, 788)  (220, 220) 0(15, 15) ,(11, 11)  "(330, 330) y(88, 88) ":(788, 788)  (220, 220) 1(16, 16) ,(11, 11)  "(330, 330) theta(15976, 15976) ":(788, 788)  null(845, 845) ,(11, 11)  "(330, 330) message(1994, 1994) ":(788, 788)  null(845, 845) }}
(11248, 11248) </tool_call>(151658, 151658) <|im_end|>(151645, 151645) 
(-100, 198)
```

* After this PR (Correct arguments for "move" tool)
``` text
(User prompt and tool tokens omitted.)
(-100, 198) <|im_start|>(-100, 151644) assistant(-100, 77091) 
(-100, 198) <think>(-100, 151667) 

(-100, 271) </think>(-100, 151668) 

(-100, 271) <tool_call>(151657, 151657) 
(198, 198) {"(4913, 4913) name(606, 606) ":(788, 788)  "(330, 330) move(3397, 3397) ",(497, 497)  "(330, 330) arguments(16370, 16370) ":(788, 788)  {"(5212, 5212) x(87, 87) ":(788, 788)  (220, 220) 0(15, 15) ,(11, 11)  "(330, 330) y(88, 88) ":(788, 788)  (220, 220) 1(16, 16) }}
(11248, 11248) </tool_call>(151658, 151658) <|im_end|>(151645, 151645) 
(-100, 198)
```

## How has this been tested?

This change was evaluated using the qwen3-1.7b model with vLLM for inference. 
The results are as follows:

Before fine-tuning: 55.5% accuracy
After fine-tuning (before this PR): 29.8% (Performance degraded)
After fine-tuning (this PR): 90.3% (Performance improved)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved prompt processing by filtering out any entries with missing values before tokenization, ensuring cleaner serialized outputs.  
* **Tests**
  * Added tests to verify that optional fields are correctly omitted in tool call outputs, preventing null values in serialized chat prompts.

### CODE REVIEW

File: src/axolotl/prompt_strategies/chat_template.py [hunk 1]
```diff
@@ -378,13 +378,21 @@ def tokenize_prompt(self, prompt: dict[str, Any]):
         """
         Public method that can handle either a single prompt or a batch of prompts.
         """
+        def _remove_none_values(obj):
+            if hasattr(obj, "items"):
+                return {k: _remove_none_values(v) for k, v in obj.items() if v is not None}
+            elif isinstance(obj, list):
+                return [_remove_none_values(elem) for elem in obj]
+            return obj
```
[the remark this answers is not in the document]
Comment (greenhestu) [reply src/axolotl/prompt_strategies/chat_template.py]: The hasattr function is necessary here because obj is not guaranteed to be a dictionary; it could be a class instance.

**Comment by winglian** [comment]
@greenhestu Thanks for tracking this down. I wrote up a quick test case, but it doesn't pass. see https://gist.github.com/winglian/b1b658a37b2dc337448aef9f10080095
 
 EDIT: found the issue and added separate tests for batched and non-batched parsing

**Review by winglian — APPROVED** [review]

**Review by NanoCode012 — APPROVED** [review]
Thank you for the clear issue and fix.

This brings up a second issue, whether this could also occur for `tools`, since it has a lot of nested elements. For example, one `int` field could have a `min` attribute. Would this also appear as `null` for a `string` field due to the potential schema merge?

Either way, your fix should catch it.

**Comment by greenhestu** [comment]
That’s definitely an issue—thanks for pointing it out.

Then we need to use something other than `load_dataset` to load the JSONL file.
* (As lhoestq mentioned in [this discussion](https://discuss.huggingface.co/t/using-datasets-to-open-jsonl/161037/10), 🤗 Datasets isn’t suitable for such a dataset when the same field can appear with different data types)

I considered using `orjsonl`, but I’m not familiar with this repo, and the loading code is heavily tied to `datasets`. Fixing it would take more time than I can spare right now. 

Would it be better to close this PR and continue the discussion in an issue instead?

**Comment by greenhestu** [comment]
Oh, it’s already been merged. Thanks!

**Comment by NanoCode012** [comment]
For now, I think your fix covers that case too. Thank you for the PR!

**Comment by greenhestu** [comment]
I now see you were talking about schema merging for nested elements.

I had confused it with a type conflict issue, like when a field type changes from a `integer` to a `string`.

``` json
{"messages":[{"role":"user","content":"send data1 1"},{"role":"assistant","content":"","tool_calls":[{"function":{"name":"send_data1","arguments":{"data":1}}}]}],"tools":[{"type":"function","function":{"name":"send_data1","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"integer","description":"the data"}},"required":["data"]}}},{"type":"function","function":{"name":"send_data2","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"string","description":"the data"}},"required":["data"]}}}],"add_generation_prompt":false}
{"messages":[{"role":"user","content":"send data2 2"},{"role":"assistant","content":"","tool_calls":[{"function":{"name":"send_data2","arguments":{"data":"2"}}}]}],"tools":[{"type":"function","function":{"name":"send_data1","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"integer","description":"the data"}},"required":["data"]}}},{"type":"function","function":{"name":"send_data2","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"string","description":"the data"}},"required":["data"]}}}],"add_generation_prompt":false}

```
> pyarrow.lib.ArrowInvalid: JSON parse error: Column(/messages/[]/tool_calls/[]/function/arguments/data) changed from number to string in row 1

This could be avoided by using different field names for different types, though.

Appreciate the review!

**Comment by gamersover** [comment]
I have a pr to solve this issue. The root cause is that arguments is of type dict. The ideal solution would be to change arguments to a string type, which would resolve all related problems. https://github.com/axolotl-ai-cloud/axolotl/pull/3136

> I now see you were talking about schema merging for nested elements.
> 
> I had confused it with a type conflict issue, like when a field type changes from a `integer` to a `string`.
> 
> ```json
> {"messages":[{"role":"user","content":"send data1 1"},{"role":"assistant","content":"","tool_calls":[{"function":{"name":"send_data1","arguments":{"data":1}}}]}],"tools":[{"type":"function","function":{"name":"send_data1","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"integer","description":"the data"}},"required":["data"]}}},{"type":"function","function":{"name":"send_data2","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"string","description":"the data"}},"required":["data"]}}}],"add_generation_prompt":false}
> {"messages":[{"role":"user","content":"send data2 2"},{"role":"assistant","content":"","tool_calls":[{"function":{"name":"send_data2","arguments":{"data":"2"}}}]}],"tools":[{"type":"function","function":{"name":"send_data1","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"integer","description":"the data"}},"required":["data"]}}},{"type":"function","function":{"name":"send_data2","description":"send data","parameters":{"type":"object","properties":{"data":{"type":"string","description":"the data"}},"required":["data"]}}}],"add_generation_prompt":false}
> ```
> 
> > pyarrow.lib.ArrowInvalid: JSON parse error: Column(/messages/[]/tool_calls/[]/function/arguments/data) changed from number to string in row 1
> 
> This could be avoided by using different field names for different types, though.
> 
> Appreciate the review!
